core/src/main/java/org/elasticsearch/action/support/nodes/BaseNodesResponse.java

```diff
@@ -91,11 +105,36 @@ public TNodeResponse getAt(int position) {
     public void readFrom(StreamInput in) throws IOException {
         super.readFrom(in);
         clusterName = ClusterName.readClusterName(in);
+        failures = in.readArray(FailedNodeException[]::new, FailedNodeException::new);
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
         clusterName.writeTo(out);
+        out.writeArray(failures);
     }
+
+    /**
+     * Expected to be invoked by any {@code BaseNodesResponse} that implements {@code ToXContent} at the top level of the response object.
+     */
+    protected XContentBuilder toInnerXContent(XContentBuilder builder, Params params) throws IOException {
```

Comment:
I don't like super methods that when they *are* overridden, they *must* be invoked. I also don't like that this method is on the base class and thus inherited by response objects that do not implement `ToXContent`, I think that is misleading. Lastly, I think that we should include total/successful/failed counts in the response.

My specific suggestion is to add a method to `RestActions` similar to `RestActions#buildBroadcastShardsHeader`. I think it can look something like this:

```java
    public static <TNodesResponse extends BaseNodeResponse> void buildNodesInfoHeader(
            final XContentBuilder builder,
            final ToXContent.Params params,
            final BaseNodesResponse<TNodesResponse> response) throws IOException {
        final TNodesResponse[] responses = response.getNodes();
        final FailedNodeException[] failures = response.failures();
        final int successful = responses != null ? responses.length : 0;
        final int failed = failures != null ? responses.length : 0;
        buildNodesInfoHeader(builder, params, successful + failed, successful, failed, failures);
    }

    public static void buildNodesInfoHeader(
            final XContentBuilder builder,
            final ToXContent.Params params,
            final int total,
            final int successful,
            final int failed,
            final FailedNodeException[] failedNodeExceptions) throws IOException {
        // nocommit: add a constant to Fields
        builder.startObject("_nodes");
        builder.field(Fields.TOTAL, total);
        builder.field(Fields.SUCCESSFUL, successful);
        builder.field(Fields.FAILED, failed);
        if (failedNodeExceptions != null && failedNodeExceptions.length > 0) {
            builder.startArray(Fields.FAILURES);
            for (FailedNodeException failedNodeException : failedNodeExceptions) {
                builder.startObject();
                failedNodeException.toXContent(builder, params);
                builder.endObject();
            }
            builder.endArray();
        }
        builder.endObject();
    }

    public static <TNodesResponse extends BaseNodesResponse & ToXContent> BytesRestResponse nodesInfoResponse(
            final XContentBuilder builder,
            final ToXContent.Params request,
            final TNodesResponse response) throws IOException {
        builder.startObject();
        RestActions.buildNodesInfoHeader(builder, request, response);
        response.toXContent(builder, request);
        builder.endObject();
        return new BytesRestResponse(RestStatus.OK, builder);
    }
```

And then `buildResponse` call sites can just look like `return RestActions.nodesInfoResponse(builder, request, response);`